File: cli/python/agent_disaster_lab/enforce.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Any

_SECTOR_MAP: dict[str, Any] = {}

def _get_sector(class_name: str) -> Any:
    if class_name not in _SECTOR_MAP:
        if class_name == "NigeriaFintechCompliance":
            from comply54 import NigeriaFintechCompliance
            _SECTOR_MAP[class_name] = NigeriaFintechCompliance()
        elif class_name == "NigeriaHealthcareCompliance":
            from comply54.sectors.nigeria_health import NigeriaHealthcareCompliance
            _SECTOR_MAP[class_name] = NigeriaHealthcareCompliance()
        elif class_name == "NigeriaInsuranceCompliance":
            from comply54.sectors.nigeria_insurance import NigeriaInsuranceCompliance
            _SECTOR_MAP[class_name] = NigeriaInsuranceCompliance()
        elif class_name == "KenyaFintechCompliance":
            from comply54 import KenyaFintechCompliance
            _SECTOR_MAP[class_name] = KenyaFintechCompliance()
        elif class_name == "PanAfricanFintechCompliance":
            from comply54 import PanAfricanFintechCompliance
            _SECTOR_MAP[class_name] = PanAfricanFintechCompliance()
        else:
            raise ValueError(f"Unknown sector class: {class_name}")
    return _SECTOR_MAP[class_name]


@dataclass
class EnforcementResult:
    blocked: bool
    overall: str
    violation: str = ""
    regulation: str = ""
    pack: str = ""


def enforce(
    action: str,
    params: dict[str, Any],
    sector_class: str,
    context: dict[str, Any] | None = None,
) -> EnforcementResult:
    try:
        compliance = _get_sector(sector_class)
        result = compliance.check(action, params, "", context or {})

        if result.blocked:
            deny = next((d for d in result.decisions if d.action == "deny"), None)
            return EnforcementResult(
                blocked=True,
                overall=result.overall,
                violation=deny.messages[0] if deny and deny.messages else "Policy violation",
                regulation=deny.regulation if deny else "",
                pack=deny.pack if deny else "",
            )
        return EnforcementResult(blocked=False, overall=result.overall)
    except Exception:
        return EnforcementResult(blocked=False, overall="allow")

File: cli/node/src/enforce.ts
import {
  NigeriaFintechCompliance,
  NigeriaHealthcareCompliance,
  NigeriaInsuranceCompliance,
  KenyaFintechCompliance,
  PanAfricanFintechCompliance,
} from "@comply54/core"

type SectorClass =
  | "NigeriaFintechCompliance"
  | "NigeriaHealthcareCompliance"
  | "NigeriaInsuranceCompliance"
  | "KenyaFintechCompliance"
  | "PanAfricanFintechCompliance"

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const SECTOR_MAP: Record<SectorClass, () => any> = {
  NigeriaFintechCompliance: () => new NigeriaFintechCompliance(),
  NigeriaHealthcareCompliance: () => new NigeriaHealthcareCompliance(),
  NigeriaInsuranceCompliance: () => new NigeriaInsuranceCompliance(),
  KenyaFintechCompliance: () => new KenyaFintechCompliance(),
  PanAfricanFintechCompliance: () => new PanAfricanFintechCompliance(),
}

export interface EnforcementResult {
  blocked: boolean
  overall: string
  violation?: string
  regulation?: string
  pack?: string
  penalty?: string
}

export function enforce(
  action: string,
  params: Record<string, unknown>,
  sectorClass: string,
  context: Record<string, unknown> = {}
): EnforcementResult {
  try {
    const factory = SECTOR_MAP[sectorClass as SectorClass]
    if (!factory) throw new Error(`Unknown sector class: ${sectorClass}`)
    const compliance = factory()
    const result = compliance.check(action, params, "", context)

    if (result.blocked) {
      const denyDecision = result.decisions?.find((d: { action: string }) => d.action === "deny")
      return {
        blocked: true,
        overall: result.overall,
        violation: denyDecision?.messages?.[0] ?? "Policy violation",
        regulation: denyDecision?.regulation ?? "Unknown regulation",
        pack: denyDecision?.pack ?? "",
      }
    }
    return { blocked: false, overall: result.overall }
  } catch {
    return { blocked: false, overall: "allow" }
  }
}
